### PR TITLE
Send build failure notifications to build definition `.yml` file synthetic sibling CODEOWNERS owners, instead of the owners of the `.yml` file itself.

### PR DESCRIPTION
**Pull request by konrad-jamrozik**
This PR implements capability as described in:
- #5181

plus does a little bit of related message output changes and code refactoring.

Copy-pasting here the relevant paragraph for convenience:

> We will make the `notification-configuration` tool search for owners not of the pipeline definition file it originated from (e.g. `/foo/ci.yml`), but a made-up sibling file named e.g. `PipelineOwner`, e.g. `/foo/PipelineOwner`. This way if there is a `CODEOWNERS` path that matches to `/foo/ci.yml` it is not going to be used to determine where to route build failure notifications, but as a `CODEOWNERS` path that matches `/foo/PipelineOwner`, e.g. `/foo/`, will. You can find the details of this approach, and preceding discussion, here:
> - https://github.com/Azure/azure-sdk-for-net/pull/33595#issuecomment-1401349175

## Testing done

I did some manual testing with C# snippets to confirm the introduced method `GetSiblingFilePath` behaves as expected. All other changes are done via automated IDE refactorings so should be safe.

## Expected behavior change

Once this change goes in effect via PR similar to this one:
- #5177

I expect there will be some changes to whom the build failure notifications are routed.

I reviewed the CODEOWNERS files targeted by the [`automation - build-failure-notification-subscriptions`](https://dev.azure.com/azure-sdk/internal/_build?definitionId=679) pipeline. Below I list the files and cases where owners who receive build failure notifications will change.

| File | Expected changes |
|-|-|
https://github.com/Azure/azure-dev/blob/main/.github/CODEOWNERS | None
https://github.com/Azure/azure-sdk-for-c/blob/main/.github/CODEOWNERS | None
https://github.com/Azure/azure-sdk-for-cpp/blob/main/.github/CODEOWNERS | None
https://github.com/Azure/azure-sdk-for-go/blob/main/.github/CODEOWNERS | None
https://github.com/Azure/azure-sdk-for-java/blob/main/.github/CODEOWNERS | [aggregate-reports.yml](https://github.com/Azure/azure-sdk-for-java/blob/0e4db4b37a923332299e4732aaa2a8c04d300ce0/.github/CODEOWNERS#L774), [build](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2133986&view=logs&j=1f7b91b6-9a2a-5c83-7d3e-ca0fa5848c83&t=cdd04468-e304-5da0-e0b2-ea410a114c81&l=1015)
https://github.com/Azure/azure-sdk-for-js/blob/main/.github/CODEOWNERS | [Over 100 .yml files](https://github.com/Azure/azure-sdk-for-js/blob/a65c4761bd0a0310607bcba151f9a037e39a76d2/.github/CODEOWNERS#L184)
https://github.com/Azure/azure-sdk-for-net/blob/main/.github/CODEOWNERS | [4 .yml files](https://github.com/Azure/azure-sdk-for-net/blob/main/.github/CODEOWNERS#L470)
https://github.com/Azure/azure-sdk-for-python/blob/main/.github/CODEOWNERS | [7 .yml files](https://github.com/Azure/azure-sdk-for-python/blob/main/.github/CODEOWNERS#L165)

Most likely we want to modify the CODEOWNERS files before applying the effects of this PR, to keep the recipients of build failure notifications unchanged. This can be achieved by adding relevant paths to the CODEOWNERS file. 

E.g. a block like:

```
# PRLabel: %Mgmt
/sdk/appplatform/arm-appplatform @qiaozha @dw511214992 @MaryGao
/sdk/appplatform/ci.mgmt.yml @qiaozha @dw511214992 @MaryGao
```

Can be replaced with:

```
/sdk/appplatform/ @qiaozha @dw511214992 @MaryGao
# PRLabel: %Mgmt
/sdk/appplatform/arm-appplatform @qiaozha @dw511214992 @MaryGao
/sdk/appplatform/ci.mgmt.yml @qiaozha @dw511214992 @MaryGao
```

## Additional design notes

Currently I assume the proposed `PipelineOwnerSyntheticFileName` is not present in the repository nor in `CODEOWNERS` file. But we might actually want to consider adding this file path to `CODEOWNERS` a valid approach, to fine-tune who gets the build notifications. Then the example above would become, for example:

```
# PRLabel: %Mgmt
/sdk/appplatform/arm-appplatform @qiaozha @dw511214992 @MaryGao
/sdk/appplatform/__PipelineOwner__ @qiaozha @dw511214992 @MaryGao
/sdk/appplatform/ci.mgmt.yml @qiaozha @dw511214992 @MaryGao
```

However, in such case, we would need to add special case that such file should **not** actually exist, in our validation spec:
- https://github.com/Azure/azure-sdk-tools/issues/4859#issuecomment-1370360622

### CODE REVIEW

File: tools/notification-configuration/notification-creator/NotificationConfigurator.cs [hunk 1]
```diff
@@ -275,6 +292,39 @@ private async Task SyncTeamWithCodeOwnerFile(BuildDefinition pipeline, WebApiTea
             }
         }
 
+        private CodeownersEntry GetMatchingCodeownersEntry(YamlProcess process, List<CodeownersEntry> codeownersEntries)
+        {
+            // See comment on PipelineOwnerSyntheticFileName to understand why the pipelineOwnerFile
+            // is not simply process.YamlFilename.
+            string pipelineOwnerFile = GetSiblingFilePath(
+                targetPath: process.YamlFilename,
+                siblingFileName: PipelineOwnerSyntheticFileName);
```
Comment (konrad-jamrozik) [inline tools/notification-configuration/notification-creator/NotificationConfigurator.cs:301]: FYI this `GetSiblingFilePath` invocation is the actual behavior change. Previously the code was equivalent to:

`string pipelineOwnerFile = process.YamlFilename;`

**Comment by konrad-jamrozik** [comment]
@weshaggard I need your input before I can proceed with this PR. Please see the description, sections "Expected behavior change" and "Additional design notes". 

Should I go ahead with adjusting all the language repo CODEOWNERS files to ensure this PR doesn't change any recipients of build failure notifications?

**Review by weshaggard — CHANGES_REQUESTED** [review]
As I mentioned in the original PR lets not go after this right now. @benbp will look into other workflows for monitoring all the yml files in the repo.

**Comment by konrad-jamrozik** [comment]
Closing this PR per Wes' comment above. Instead, I submitted a PR that incorporates refactorings made in this PR, and more:
- #5214

**Pull request closed without merging**